### PR TITLE
Issue-1668: ChannelBindException detection fails

### PR DESCRIPTION
**Pull request by roggenbrot**
Original issue: #1668

This PR fixes the ChannelBindException detection by using a non i18n part of the exception message

### CODE REVIEW

**Review by violetagg — APPROVED** [review]

**Comment by violetagg** [comment]
@reactor/netty-team PTAL

**Review by simonbasle — CHANGES_REQUESTED** [review]
that part of the code should be unit-testable. the main challenge being the `Supplier<SocketAdress>` I guess

**Comment by violetagg** [comment]
@@roggenbrot Will you be able to add a test where the exception is constructed with `bind(..) failed: Die Adresse wird bereits verwendet` for example and then we can check that we still work as expected?

**Comment by roggenbrot** [comment]
I will add a test case.

But i think the part I fixed is just copy&paste from ` ChannelBindException.fail`.

I will fix ` ChannelBindException.fail`, change NewConnectionProvider to use `ChannelBindException.fail` and add a test case for ` ChannelBindException fail`

**Review by simonbasle — APPROVED** [review]

**Comment by simonbasle** [comment]
@roggenbrot good find! The PR needs some formatting polish (see checkstyle complains) but should otherwise be good to go

File: reactor-netty-core/src/test/java/reactor/netty/ChannelBindExceptionTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	@Test
+	@SuppressWarnings("unchecked")
+	void testFailBindExceptionCause() {
+
+		UnsupportedOperationException cause = new UnsupportedOperationException();
+		ChannelBindException ex = ChannelBindException.fail(new InetSocketAddress("test", 4956), cause);
+		assertThat(ex.getCause()).isEqualTo(cause);
+		assertThat(ex.localHost()).isEqualTo("test");
+		assertThat(ex.localPort()).isEqualTo(4956);
+		assertThat(ex.localPort()).isEqualTo(4956);
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/ChannelBindExceptionTest.java:37]: nit: Duplicating row?

File: reactor-netty-core/src/test/java/reactor/netty/resources/NewConnectionProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		Mockito.verify(sink).error(Mockito.argThat(a -> a instanceof UnsupportedOperationException));
+
+		connect = new NewConnectionProvider.DisposableConnect(sink, () -> new InetSocketAddress("test", 4956));
+		// Not possible to mock io.netty.channel.unix.Errors.NativeIoException or create a new instance because of Jni errors
+		// java.lang.UnsatisfiedLinkError: 'int io.netty.channel.unix.ErrorsStaticallyReferencedJniMethods.errnoENOENT()'
+		connect.onError(new IOException("bind(..) failed: Address already in use"));
+		Mockito.verify(sink).error(Mockito.argThat(a -> a instanceof ChannelBindException && ((ChannelBindException) a).localHost().equals("test") && ((ChannelBindException) a).localPort() == 4956));
+
+		// Issue-1668
+		connect = new NewConnectionProvider.DisposableConnect(sink, () -> new InetSocketAddress("tset", 6956));
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/NewConnectionProviderTest.java:46]: nit: typos -> `tset` vs `test`

**Review by violetagg — CHANGES_REQUESTED** [review]
nit: There is no strong requirement but aim to wrap code at 120 characters

File: reactor-netty-core/src/test/java/reactor/netty/ChannelBindExceptionTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import java.io.IOException;
+import java.net.BindException;
+import java.net.InetSocketAddress;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ChannelBindExceptionTest {
+
+	@Test
+	@SuppressWarnings("unchecked")
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/ChannelBindExceptionTest.java:29]: Do we need this suppression?

File: reactor-netty-core/src/test/java/reactor/netty/resources/NewConnectionProviderTest.java (hunk 2 above)
Comment (roggenbrot) [reply reactor-netty-core/src/test/java/reactor/netty/resources/NewConnectionProviderTest.java:46]: This typos are by purpose to test other string on second iteration. I will change it to test1..

Comment (violetagg) [reply reactor-netty-core/src/test/java/reactor/netty/resources/NewConnectionProviderTest.java:46]: thanks

**Review by simonbasle — APPROVED** [review]

**Comment by simonbasle** [comment]
@roggenbrot nothing to do with the PR content itself, but for future reference having these kind of commits for a PR is not great for the review process. looking at the log:

image

The merge commits are very weird, and I would expect no rebase nor merges during the review phase of a PR unless absolutely necessary (ie. you depend on a feature that is newly introduced in the base branch).

So I would have expected multiple commits that each represent a meaningful set of changes, ie. addressing review comments made since the last commit - and only that.
I would also expect commit message titles that are more descriptive (rather than a repetition of the first commit message).

Something like:

```
original commit title
review: add a test simulating german in exception message
review: add since tag to javadoc
review: fix formatting
review: fix formatting again 
```

The pattern of commits in this PR blurs what changed when coming back to the PR for review. fortunately it is small in size and can be compared as a whole to `main`, but had it been larger, it would have been very hard to review incrementally.

**Comment by roggenbrot** [comment]
@simonbasle Sorry, struggling/fighting a little bit with amend/IntelliJ/gitflow which I'm not used to. I'm trying to do it better next time. Thanks for feedback

**Review by violetagg — APPROVED** [review]

**Pull request merged**